src/debugger/direct/directDebugAdapter.ts

```diff
@@ -0,0 +1,192 @@
+    private isSettingsInitialized: boolean; // used to prevent parameters reinitialization when attach is called from launch
+    private previousAttachArgs: IDirectAttachRequestArgs;
+
+    public constructor(opts: IChromeDebugSessionOpts, debugSession: ChromeDebugSession) {
+        super(opts, debugSession);
+        this.outputLogger = (message: string, error?: boolean | string) => {
+            let category = "console";
+            if (error === true) {
+                category = "stderr";
+            }
+            if (typeof error === "string") {
+                category = error;
+            }
+
+            let newLine = "\n";
+            if (category === "stdout" || category === "stderr") {
+                newLine = "";
+            }
+            debugSession.sendEvent(new OutputEvent(message + newLine, category));
+        };
+
+        this.isSettingsInitialized = false;
+    }
+
+    public launch(launchArgs: IDirectLaunchRequestArgs): Promise<void>  {
+        const extProps = {
+            platform: {
+                value: launchArgs.platform,
+                isPii: false,
+            },
+            isDirect: {
+                value: true,
+                isPii: false,
+            },
+        };
+
+        return new Promise<void>((resolve, reject) => this.initializeSettings(launchArgs)
+            .then(() => {
+                this.outputLogger("Launching the app");
+                logger.verbose(`Launching the app: ${JSON.stringify(launchArgs, null , 2)}`);
+                return TelemetryHelper.generate("launch", extProps, (generator) => {
+                    return this.remoteExtension.launch({ "arguments": launchArgs })
+                        .then(() => {
+                            return this.remoteExtension.getPackagerPort(launchArgs.cwd);
+                        })
+                        .then((packagerPort: number) => {
+                            launchArgs.port = packagerPort;
+                            this.attach(launchArgs).then(() => {
+                                resolve();
+                            }).catch((e) => reject(e));
+                        }).catch((e) => reject(e));
+                })
+                .catch((err) => {
+                    this.outputLogger("An error occurred while launching the application. " + err.message || err, true);
+                    this.cleanUp();
+                    reject(err);
+                });
+        }));
+    }
+
+    public attach(attachArgs: IDirectAttachRequestArgs): Promise<void> {
+        const extProps = {
+            platform: {
+                value: attachArgs.platform,
+                isPii: false,
+            },
+            isDirect: {
+                value: true,
+                isPii: false,
+            },
+        };
+
+        this.previousAttachArgs = attachArgs;
+
+        return new Promise<void>((resolve, reject) => this.initializeSettings(attachArgs)
```

Comment:
We used promises because the indirect debugger use such approach. We can try to move to async/awaits, but we need to refactor a lot of code and tests to work with them. Should we add this task to backlog?

Reply:
Yes. Lets open an issue on our backlog.  Thanks.